### DISCUSSION TITLE
runtimeCaching on nextjs /api paths

### DISCUSSION THREAD

**Discussion by bart-sofomo** [serwist/serwist#160] [Q&A]
Hi,
I wrote an app that is using nextjs and wundergraph. The wundergraph is served under `/api/wg` route so that I can use various hooks provider from wundergraph to get the data on client side.
The problem is that when I leave the app on the phone and go back to it after like 5minutes all the places in my page where I use wundergraph on the client side gives me errors with 401 unauthorized.

And here comes my confusion - the `/api` paths are cached by default in serwist - am I right?
Wundergraph requires authentication (we are using Clerk with nextjs and wundergraph with jwks url) and basically my data under `/api/wg` can be really dynamic which means if I go out from the page and go back to it - it should fetch data again.
Am I thinking correctly that I should do something like:
```
  runtimeCaching: [
    {
      matcher: ({ url }) => url.pathname.startsWith('/api/wg/'),
      handler: new NetworkOnly(),
    },
    ...defaultCache,
  ],
```
in order to achieve always fresh data from `/api/wg` then I go back to the page (hooks are calling data automatically). 
Maybe it's something more connected with authorization with clerk? Any help will be appreciated :)

**Comment by DuCanhGH** [comment]
@bart-sofomo

> The problem is that when I leave the app on the phone and go back to it after like 5minutes all the places in my page where I use wundergraph on the client side gives me errors with 401 unauthorized.

That's... weird. `@serwist/next` uses `NetworkFirst`, which does not use the cache while you are online, by default for API routes. Not sure how this issue would be on Serwist.

> in order to achieve always fresh data from `/api/wg`

Well, yeah. That forces Serwist to never ever cache the data.

> Maybe it's something more connected with authorization with clerk? Any help will be appreciated :)

I have never used Clerk and, as a result, don't know much about it, but I'd say that things should work just fine. Well, do check if anything related to authn/authz is stuck in the cache :) For example, HTTP headers often go under the radar, and them being stale often causes issues with authn/authz.

**Reply by bart-sofomo** [reply]
> check if anything related to authn/authz is stuck in the cache

* What do you mean by that? I mean where should I look? 

* To more describe clerk:
It uses middleware (https://clerk.com/docs/references/nextjs/clerk-middleware) in which you can define which routes should be protected. In case or wundergraph we are doing something like this in middleware:
```
if (isWundergraphRoute(req)) {
    // get clerk token
    const { getToken } = auth();
    const token = await getToken({ template: 'wundergraph' });
    const pathname = req.nextUrl.pathname.replace('/api/wg', '');

    // rewrite the api url to the WunderGraph API
    const url = new URL(
      pathname + req.nextUrl.search,
      process.env.NEXT_WUNDERGRAPH_URL || 'http://localhost:9991',
    );

    // add the token to the Authorization header
    const headers = new Headers({
      Authorization: `Bearer ${token}`,
    });

    // rewrite the request to the WunderGraph API
    const response = NextResponse.rewrite(url, {
      request: {
        headers,
      },
    });

    return response;
  }
```
I think this is calling `api.clerk.com/v1/sessions/*`.
So basically there is a client that always points to `/api/wg` and in middleware we are attaching token. Is it possible that this token is somehow cached or maybe the call to `api.clerk.com/v1/sessions/*` is cached? 
More funny is that it is only happening in mobile browser, it does not happen in desktop :/

**Reply by DuCanhGH** [reply]
@bart-sofomo 

> What do you mean by that? I mean where should I look?

Any cached response may contain stale data. This stale data may be in the response's URL, headers, or body. I'd use `CacheableResponse` with a list of header names that Clerk, Wundergraph, and whatnot may use for authentication or authorization. But eh, let's just forget this one. I don't think there's anything in the response. Most of the cases...

> It uses middleware in which you can define which routes should be protected.

Well, the middleware is bypassed whenever a cached response is used. You know, that's because the request does not go back to the server. Basically, if your request to "/api/wg" is the same as any older request that corresponds to a cached response, that response is immediately used, and no actual network request is made. So you may want to use `NetworkOnly`/`NetworkFirst` for any protected resource.

> I think this is calling api.clerk.com/v1/sessions/*.
So basically there is a client that always points to /api/wg and in middleware we are attaching token. Is it possible that this token is somehow cached or maybe the call to api.clerk.com/v1/sessions/* is cached?

The service worker does not affect whatever that may happen on the server by any means. This means that no, the token should not be stale, and the call to Clerk's API, well, may be cached, but not by your service worker.

**Comment by bart-sofomo** [comment]
@DuCanhGH problem seems to be on the clerk/nextjs side - I disabled serwist and I'm running into exactly the same issue. Sorry to bother 🙂

**Reply by DuCanhGH** [reply]
@bart-sofomo no worries! I'd clear the caches just to make sure that this is not on Serwist though :)
